content/articles/building-spotify-recommendation-engine/index.md

```diff
@@ -0,0 +1,349 @@
+    
+    playlist_feature_set_weighted_final = playlist_feature_set_weighted.iloc[:, :-3]
+    
+
+    
+    return playlist_feature_set_weighted_final.sum(axis = 0), spotify_features_nonplaylist
+
+```
+```python
+playlist_vector, nonplaylist_df = generate_playlist_vector(spotify_features_df, playlist_df, 1.2)
+print(playlist_vector.shape)
+print(nonplaylist_df.head())
+```
+![snapshot](/engineering-education/building-spotify-recommendation-engine/snapshot-5.jpg)
+
+#### Generating Recommendations
+
+Now comes the final part of our implementation, generating recommendations. As stated before, we are going to use cosine similarity as a similarity metric to determine the songs that are very much similar to our playlist. 
+
+We will perform the cosine similarity between our playlist vector and songs not present in our playlist. 
+
+First, we will perform cosine similarity using a python-based library  Scikit and store the cosine similarity values in a separate column. Later we will reverse sort the dataframe based on the cosine similarity column. Finally, we will generate the top 15 recommendations of songs as our recommended playlist.
+
+```python
+def generate_recommendation(spotify_data, playlist_vector, nonplaylist_df):
+
+    non_playlist = spotify_data[spotify_data['track_id'].isin(nonplaylist_df['track_id'].values)]
+    non_playlist['sim'] = cosine_similarity(nonplaylist_df.drop(['track_id'], axis = 1).values, playlist_vector.drop(labels = 'track_id').values.reshape(1, -1))[:,0]
+    non_playlist_top15 = non_playlist.sort_values('sim',ascending = False).head(15)
+    non_playlist_top15['url'] = non_playlist_top15['track_id'].apply(lambda x: sp.track(x)['album']['images'][1]['url'])
+    
+    return  non_playlist_top15
+```
+```python
+top15 = generate_recommendation(spotify_data, playlist_vector, nonplaylist_df)  
+top15.head()
+```
+![snapshot](/engineering-education/building-spotify-recommendation-engine/snapshot-6.jpg)
+
+```python
+#Visulaizing the cover-art of the recommended playlist
+visualize_cover_art(top10)
+```
+![snapshot](/engineering-education/building-spotify-recommendation-engine/recommendation.png)
+
+### Conclusion
+
+Using the concept of cosine similarity, we have seen that we were able to generate recommendations of songs that are similar to our playlist. But we have to note that Recommendation Systems are not perfect every time. As in this case, we have used a Content-Based Filtering mechanism. The model can only make recommendations based on the user's existing interests. Thus, we have limited the ability of a user to expand their existing interests. One more drawback of this implementation is that the Spotify songs dataset that we have downloaded from the kaggle dataset doesn't necessarily contain all the songs in your Spotify playlist. Therefore, you will not be able to utilize your playlist fully to generate recommendations.
```

Comment:
Add a link to the colab notebook containing the entire code. This will ensure reproducibility of the code.